File: Math & Geometry/0974. Subarray Sums Divisible by K/0974.py
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        remainder_count = defaultdict(int)
        remainder_count[0] = 1

        total = 0
        res = 0

        for i in range(len(nums)):
            total += nums[i]
            remainder = total % k
            if remainder < 0:
                remainder += k
            
            res += remainder_count[remainder]
            remainder_count[remainder] += 1

        return res

File: Math & Geometry/0974. Subarray Sums Divisible by K/0974.cpp
#pragma GCC optimize("O3","Oz","unroll-loops","inline-functions","fast-math","march=native","lto","omit-frame-pointer")
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        int32_t prefix = 0;
        int32_t res = 0;
        std::unordered_map<int32_t, int32_t> seenBefore;
        seenBefore[0] = 1;

        for (auto& num : nums) {
            prefix += num;
            prefix %= k;
            if (prefix < 0) prefix += k;
            res += seenBefore[prefix];
            seenBefore[prefix]++;
        }
        return res;
    }
};